src/syscall_sdl.c

```diff
@@ -276,3 +337,162 @@ void syscall_submit_queue(riscv_t *rv)
+
+    memcpy(sfx_samples, ptr, sizeof(uint8_t) * sfx_sample_nums);
+    sfx_chunk = Mix_QuickLoad_RAW(sfx_samples, sfx_sample_nums);
+    if (!sfx_chunk)
+        return NULL;
+
+    chan = Mix_PlayChannel(-1, sfx_chunk, 0);
+    if (chan == -1)
+        return NULL;
+
+    /* multiplied by 8 because sfx->volume's max is 15 */
+    Mix_Volume(chan, sfx->volume * 8);
+
+    Mix_Playing(chan);
+
+    return NULL;
+}
+
+void *music_handler(void *arg)
+{
+    sound_t *music = (sound_t *) arg;
+    int looping = music->looping ? -1 : 1;
+
+    if (mid_data)
+        free(mid_data);
+
+    mid_data = mus2midi(music->data, (int *) &music->size);
+    if (!mid_data) {
+        fprintf(stderr, "mus2midi failed\n");
+        return NULL;
+    }
+
+    SDL_RWops *rwops = SDL_RWFromMem(mid_data, music->size);
+    mid = Mix_LoadMUSType_RW(rwops, MUS_MID, 1);
+
+    /*
+     * multiplied by 8 because sfx->volume's max is 15
+     * further setting volume via syscall_set_music_volume
+     */
+    Mix_VolumeMusic(music->volume * 8);
+
+    if (Mix_PlayMusic(mid, looping) == -1) {
+        fprintf(stderr, "Mix_PlayMusic failed: %s\n", Mix_GetError());
+        return NULL;
+    }
+
+    return NULL;
+}
+
+void syscall_play_sfx(riscv_t *rv)
+{
+    state_t *s = rv_userdata(rv); /* access userdata */
+
+    const uint32_t sfxinfo_addr = (uint32_t) rv_get_reg(rv, rv_reg_a0);
+    int volume = rv_get_reg(rv, rv_reg_a1);
+
+    sfxinfo_t sfxinfo;
+    memory_read(s->mem, (uint8_t *) &sfxinfo, sfxinfo_addr, sizeof(sfxinfo_t));
+
+    /*
+     * hack of offset as DOOM is 32-bit but
+     * Emulator is 64-bit when accessing the sfxinfo_t
+     */
+    uint32_t sfx_data_offset = *(uint32_t *) ((uint8_t *) &sfxinfo + 24);
+    uint32_t sfx_data_size = *(uint32_t *) ((uint8_t *) &sfxinfo + 28);
+    uint8_t sfx_data[SFX_SAMPLE_SIZE];
+    memory_read(s->mem, sfx_data, sfx_data_offset,
+                sizeof(uint8_t) * sfx_data_size);
+
+    sound_t sfx = {
+        .data = sfx_data,
+        .size = sfx_data_size,
+        .volume = volume,
+    };
+    pthread_create(&sfx_thr, NULL, sfx_handler, &sfx);
+}
+
+void syscall_set_music_volume(riscv_t *rv)
+{
+    int volume = rv_get_reg(rv, rv_reg_a0);
+
+    /* multiplied by 8 because volume's max is 15 */
+    Mix_VolumeMusic(volume * 8);
+}
+
+void syscall_stop_music(riscv_t *rv UNUSED)
+{
+    if (Mix_PlayingMusic())
+        Mix_HaltMusic();
+}
+
+void syscall_play_music(riscv_t *rv)
+{
+    state_t *s = rv_userdata(rv); /* access userdata */
+
+    const uint32_t musicinfo_addr = (uint32_t) rv_get_reg(rv, rv_reg_a0);
+    int looping = rv_get_reg(rv, rv_reg_a1);
+    int volume = rv_get_reg(rv, rv_reg_a2);
+
+    musicinfo_t m;
+    memory_read(s->mem, (uint8_t *) &m, musicinfo_addr, sizeof(musicinfo_t));
+
+    /*
+     * hack of offset as DOOM is 32-bit but
+     * Emulator is 64-bit when accessing the musicinfo_t
+     */
+    uint32_t m_data_offset = *(uint32_t *) ((uint8_t *) &m + 8);
```

Comment:
The prefix `m_` is useless. Older compilers and development environments were stupid, and old programmers manually made use of such hacky prefix. Now, drop them.